jenkins-plugin/src/main/java/org/jenkinsci/plugins/pipeline/maven/publishers/JUnitUtils.java

```diff
@@ -0,0 +1,108 @@
+            throws IOException, InterruptedException {
+        TaskListener listener = context.get(TaskListener.class);
+        FilePath workspace = context.get(FilePath.class);
+        Run run = context.get(Run.class);
+        Launcher launcher = context.get(Launcher.class);
+        try {
+            // see hudson.tasks.junit.pipeline.JUnitResultsStepExecution.run
+            FlowNode node = context.get(FlowNode.class);
+            String nodeId = node.getId();
+            List<FlowNode> enclosingBlocks = JUnitResultsStepExecution.getEnclosingStagesAndParallels(node);
+            PipelineTestDetails pipelineTestDetails = new PipelineTestDetails();
+            pipelineTestDetails.setNodeId(nodeId);
+            pipelineTestDetails.setEnclosingBlocks(JUnitResultsStepExecution.getEnclosingBlockIds(enclosingBlocks));
+            pipelineTestDetails.setEnclosingBlockNames(JUnitResultsStepExecution.getEnclosingBlockNames(enclosingBlocks));
+
+            if (LOGGER.isLoggable(Level.FINER)) {
+                listener.getLogger().println("[withMaven] " + publisherName + " - collect test reports: testResults=" + archiver.getTestResults()
+                        + ", healthScaleFactor=" + archiver.getHealthScaleFactor());
+            }
+            TestResultSummary testResultSummary = JUnitResultArchiver.parseAndSummarize(archiver, pipelineTestDetails, run, workspace, launcher, listener);
+
+            if (testResultSummary == null) {
+                // no unit test results found
+                if (LOGGER.isLoggable(Level.FINE)) {
+                    listener.getLogger().println("[withMaven] " + publisherName + " - no unit test results found, ignore");
+                }
+            } else if (testResultSummary.getFailCount() == 0) {
+                // unit tests are all successful
+                if (LOGGER.isLoggable(Level.FINE)) {
+                    listener.getLogger().println("[withMaven] " + publisherName + " - unit tests are all successful");
+                }
+            } else {
+                if (LOGGER.isLoggable(Level.FINE)) {
+                    listener.getLogger().println(
+                            "[withMaven] " + publisherName + " - " + testResultSummary.getFailCount() + " unit test failure(s) found, mark job as unstable");
+                }
+                node.addAction(new WarningAction(Result.UNSTABLE).withMessage(testResultSummary.getFailCount() + " unit test failure(s) found"));
+                run.setResult(Result.UNSTABLE);
+            }
+        } catch (RuntimeException e) {
+            listener.error("[withMaven] " + publisherName + " - exception archiving JUnit results " + testResults + ": " + e + ". Failing the build.");
+            LOGGER.log(Level.WARNING, "Exception processing " + testResults, e);
+            run.setResult(Result.FAILURE);
```

Comment:
This is an antipattern. If your Pipeline step encounters a random exception, just throw it up. Do not call `setResult(FAILURE)`.

(`setResult(UNSTABLE)` as in the `junit` plugin is also somewhat problematic but there is not a great replacement.)

Reply:
you're right and I agree. For now the way the plugin is coded, it tries to be as discrete as possible to let the focus on the maven execution itself (to be clear, it catches exception everywhere). I know that you will answer that if someone use this plugin, it is its wish, one could just ` sh { mvn .... }`  and not use this plugin at all. So if we use this plugin and it has a problem, it should go to the surface to fill an issue and be fixed.

I will work on it to make all this right